### PR TITLE
Body proto

### PR DESCRIPTION
**Pull request by eggrobin**

### CODE REVIEW

File: serialization/physics.proto [hunk 1]
```diff
@@ -1,9 +1,33 @@
...
+    MasslessBody massless_body = 2;
+  }
+}
+
+message MasslessBody {
+}
+
+message MassiveBody {
+  required Quantity gravitational_parameter = 1;
+  extensions 2000 to 2999;  // Last used: TODO.
```
Comment (pleroy) [inline serialization/physics.proto]: TODO

File: serialization/physics.proto [hunk 2]
```diff
@@ -1,9 +1,33 @@
...
 
+message Body {
+  oneof body {
+    MassiveBody massive_body = 1;
+    MasslessBody massless_body = 2;
+  }
+}
+
+message MasslessBody {
+}
```
Comment (pleroy) [inline serialization/physics.proto]: Move to previous line.

File: physics/body.hpp [hunk 3]
```diff
@@ -23,6 +25,16 @@ class Body {
   template<typename Frame>
   bool is_compatible_with() const;
 
+  virtual void WriteToMessage(not_null<serialization::Body*> message) const = 0;
+
+  // Dispatches to |MassiveBody| or |MasslessBody| depending on the contents of
+  // the message.
+  // Beware the Jabberwock: if it dispatches to |OblateBody|, this method will
```
Comment (pleroy) [inline physics/body.hpp]: my son!

File: physics/body_body.hpp [hunk 4]
```diff
@@ -13,6 +15,16 @@ bool Body::is_compatible_with() const {
                              Frame::is_inertial>::is_compatible_with(this);
 }
 
+inline not_null<std::unique_ptr<Body>> Body::ReadFromMessage(
+    serialization::Body const& message) {
+  if (message.has_massless_body()) {
+    return MasslessBody::ReadFromMessage(message.massless_body());
+  } else {
```
Comment (pleroy) [inline physics/body_body.hpp]: Some preference for `else if (...has_massive...)` and then `LOG(FATAL)`.

File: physics/body_test.cpp [hunk 5]
```diff
@@ -0,0 +1,141 @@
+#include "physics/body.hpp"
+
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+
+using principia::geometry::Normalize;
+using testing::Eq;
+using testing::NotNull;
+using testing::IsNull;
```
Comment (pleroy) [inline physics/body_test.cpp]: I cannot alphabet.

File: physics/body_test.cpp [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+}
+
+// The best serialization revenge.
+TEST_F(BodyTest, MassiveSerializationSuccess) {
+  serialization::Body message;
+  MassiveBody const* cast_massive_body;
+  massive_body_.WriteToMessage(&message);
+  EXPECT_TRUE(message.has_massive_body());
+  EXPECT_FALSE(message.has_massless_body());
+  EXPECT_THAT(message.massive_body().gravitational_parameter().magnitude(),
```
Comment (pleroy) [inline physics/body_test.cpp]: EXPECT_EQ?

File: physics/body_test.cpp [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+  massless_body_.WriteToMessage(&message);
+  EXPECT_TRUE(message.has_massless_body());
+  EXPECT_FALSE(message.has_massive_body());
+
+  // Direct deserialization.
+  // No members to test in this class, we just check that it doesn't crash.
+  massless_body_ = *MasslessBody::ReadFromMessage(message);
+
+  // Dispatching from |Body|.
+  not_null<std::unique_ptr<Body>> body = Body::ReadFromMessage(message);
```
Comment (pleroy) [inline physics/body_test.cpp]: const missing here and in many places.

File: physics/frame.hpp [hunk 8]
```diff
@@ -16,6 +16,14 @@ class Frame {
   Frame() = delete;
 };
 
+enum UnknownTag {
+  kUnchecked
+};
+
+// This frame should be used for objects whose reference frame cannot be known
+// at compile time.
+using UnknownInertialFrame = Frame<UnknownTag, kUnchecked, true>;
```
Comment (pleroy) [inline physics/frame.hpp]: Comment the `true`.

File: physics/massive_body.hpp [hunk 9]
```diff
@@ -1,4 +1,11 @@
...
+﻿// The files containing the tree of of child classes of |Body| must be included
+// in the order of inheritance to avoid circular dependencies.  This class will
+// end up being reincluded as part of the implementation of its parent.
+#ifndef PRINCIPIA_PHYSICS_BODY_HPP_
+#include "physics/body.hpp"
+#else
+#ifndef PRINCIPIA_PHYSICS_MASSIVE_BODY_HPP_
+#define PRINCIPIA_PHYSICS_MASSIVE_BODY_HPP_
 
 #include "physics/body.hpp"
```
Comment (pleroy) [inline physics/massive_body.hpp]: Remove, probably.

File: physics/massive_body.hpp [hunk 10]
```diff
@@ -29,6 +36,22 @@ class MassiveBody : public Body {
...
+  void WriteToMessage(not_null<serialization::Body*> message) const override;
+
+  virtual void WriteToMessage(
+      not_null<serialization::MassiveBody*> message) const;
+
+  // Both methods below dispatch to |OblateBody<UnknownFrame>| if the
+  // |OblateBody| extension is present in the message.  Use |reinterpret_cast|
+  // afterwards as appropriate if the frame is known.
+
+  // Fails unless |message.has_massless_body()|.
```
Comment (pleroy) [inline physics/massive_body.hpp]: Do do no double negation: has_massless must be true.

File: physics/massive_body_body.hpp [hunk 11]
```diff
@@ -2,7 +2,9 @@
 
 #include "physics/massive_body.hpp"
 
+#include "physics/frame.hpp"
```
Comment (pleroy) [inline physics/massive_body_body.hpp]: I cannot alphabet.

File: physics/massive_body_body.hpp [hunk 12]
```diff
@@ -42,5 +44,33 @@ inline bool MassiveBody::is_oblate() const {
   return false;
 }
 
+inline void MassiveBody::WriteToMessage(
+    not_null<serialization::Body*> message) const {
```
Comment (pleroy) [inline physics/massive_body_body.hpp]: const all the not_null's

File: physics/massless_body.hpp [hunk 13]
```diff
@@ -1,8 +1,16 @@
...
+// end up being reincluded as part of the implementation of its parent.
+#ifndef PRINCIPIA_PHYSICS_BODY_HPP_
+#include "physics/body.hpp"
+#else
+#ifndef PRINCIPIA_PHYSICS_MASSLESS_BODY_HPP_
+#define PRINCIPIA_PHYSICS_MASSLESS_BODY_HPP_
 
 #include <vector>
 
 #include "physics/body.hpp"
```
Comment (pleroy) [inline physics/massless_body.hpp]: Remove?

File: physics/massless_body.hpp [hunk 14]
```diff
@@ -17,9 +25,24 @@ class MasslessBody : public Body {
 
   // Returns false.
   bool is_oblate() const override;
+
+  void WriteToMessage(not_null<serialization::Body*> message) const override;
+
+  virtual void WriteToMessage(
+      not_null<serialization::MasslessBody*> message) const;
+
+  // Fails unless |message.has_massless_body()|.
```
Comment (pleroy) [inline physics/massless_body.hpp]: Not entirely unlike the other comment that we never wrote.

File: serialization/physics.proto (hunk 1 above)
Comment (eggrobin) [inline serialization/physics.proto]: Done.

File: serialization/physics.proto (hunk 2 above)
Comment (eggrobin) [inline serialization/physics.proto]: Done.

File: physics/body.hpp (hunk 3 above)
Comment (eggrobin) [inline physics/body.hpp]: Done.

File: physics/body_body.hpp (hunk 4 above)
Comment (eggrobin) [inline physics/body_body.hpp]: Done.

File: physics/body_test.cpp (hunk 5 above)
Comment (eggrobin) [inline physics/body_test.cpp]: Done.

File: physics/body_test.cpp (hunk 6 above)
Comment (eggrobin) [inline physics/body_test.cpp]: Done.

File: physics/body_test.cpp (hunk 7 above)
Comment (eggrobin) [inline physics/body_test.cpp]: Done.

File: physics/frame.hpp (hunk 8 above)
Comment (eggrobin) [inline physics/frame.hpp]: Done.

File: physics/massive_body.hpp (hunk 9 above)
Comment (eggrobin) [inline physics/massive_body.hpp]: Done.

File: physics/massive_body.hpp (hunk 10 above)
Comment (eggrobin) [inline physics/massive_body.hpp]: Done.

File: physics/massive_body_body.hpp (hunk 11 above)
Comment (eggrobin) [inline physics/massive_body_body.hpp]: Done.

File: physics/massive_body_body.hpp (hunk 12 above)
Comment (eggrobin) [inline physics/massive_body_body.hpp]: Done.

File: physics/massless_body.hpp (hunk 13 above)
Comment (eggrobin) [inline physics/massless_body.hpp]: Done.

File: physics/massless_body.hpp (hunk 14 above)
Comment (eggrobin) [inline physics/massless_body.hpp]: Not undone.

**Comment by pleroy** [comment]
LGTM, but Jenkins needs to be made happy.

**Pull request merged**